tactics/hints.ml

```diff
@@ -1730,8 +1730,20 @@ let pr_hint_db_env env sigma db =
       hov 0 (goal_descr ++ str " -> " ++ hints)
     in
     let hints =
+      let str h = Pp.string_of_ppcmds (pr_global h) in
+      let regex = Str.regexp {|[^\.]+$|} in
```

Comment:
??? Don't use a regexp to parse the output of the printer, apart from being extremely contrived and probably incorrect in some fringe cases, you can get the label of the constant directly using the various `label` functions. Which also saves you from having to *print* the globref, you can just compare the output of that function.

Reply:
`h` has type `Pp.t -> Pp.t`, so I don't see how to call `label` routines.  (The match code I added is the wrong thing, just a first guess.)

> Don't use a regexp to parse the output of the printer, apart from being extremely contrived and probably incorrect

It's not contrived.  The relationship between the last component of the name and the printed value is completely obvious without looking outside this routine.  Doing a match several cases is harder to test (need to generate each type of value (often not obvious) and assure the label value matches the `pr_global` output).  What I did is a simple approach that doesn't require reverse engineering `pr_global`.  Very hard to find things in the Coq codebase.  I spend untold hours on it.

"probably incorrect" is merely FUD.  What I had should work fine even for unicode characters.

Please advise.

Reply:
`label` is not what we want for inductives and constructors
also if we want to compare the printed paths we need to go through the nametab

I guess something like
~~~ocaml
    let hints =
      let name x =  Nametab.shortest_qualid_of_global Id.Set.empty x in
      let order (h1, _, _) (h2, _, _) =
        Option.compare (fun a b ->
            let a = name a and b = name b in
            let rv = Id.compare (qualid_basename a) (qualid_basename b) in
            if rv <> 0 then rv else
              String.compare (string_of_qualid a) (string_of_qualid b))
          h1 h2
      in
~~~


>h has type Pp.t -> Pp.t, so I don't see how to call label routines

`h` in your code is Pp.h (Pp is open in this file)

Reply:
That works, thanks!  Should be ready to merge now.

Reply:
Much better indeed.